File: services/product-service/internal/handlers/health.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/mongo"
)

type HealthHandler struct {
	Mongo *mongo.Client
}

// Liveness never touches Mongo - a stalled DB should not cause k8s to
// restart otherwise-healthy pods (that would just add restart storms on
// top of a DB incident).
func (h *HealthHandler) Live(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"status": "ok"})
}

// Readiness pings Mongo with a short timeout; k8s removes this pod from the
// Service's endpoint list (but keeps it running) until it passes again.
func (h *HealthHandler) Ready(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), 2*time.Second)
	defer cancel()
	if err := h.Mongo.Ping(ctx, nil); err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"status": "not_ready"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"status": "ready"})
}

File: services/cart-service/app/routers/health.py
from fastapi import APIRouter, Depends, Response
from redis import Redis
from app.redis_client import get_redis

router = APIRouter()


# Liveness never touches Redis - only checks the process itself is running.
@router.get("/health/live")
def live():
    return {"status": "ok"}


# Readiness pings Redis; k8s pulls the pod out of Service endpoints (but
# does not restart it) until this passes again.
@router.get("/health/ready")
def ready(response: Response, r: Redis = Depends(get_redis)):
    try:
        r.ping()
        return {"status": "ready"}
    except Exception:
        response.status_code = 503
        return {"status": "not_ready"}
